### PR TITLE
IG-17095 check client and server version match

### PR DESCRIPTION
**Pull request by dinal**

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -272,6 +274,11 @@ func (api *API) History(request *frames.HistoryRequest, out chan frames.Frame) e
 	return api.historyServer.GetLogs(request, out)
 }
 
+// Exec executes a command on the backend
```
Comment (talIguaz) [inline api/api.go:277]: wrong comment

File: cmd/framesd/framesd.go [hunk 2]
```diff
@@ -88,7 +88,7 @@ func main() {
 		}
 	}
 
-	hsrv, err := framesHttp.NewServer(cfg, config.httpAddr, framesLogger, historyServer)
+	hsrv, err := framesHttp.NewServer(cfg, config.httpAddr, framesLogger, historyServer, Version)
```
Comment (talIguaz) [inline cmd/framesd/framesd.go:91]: It doesn't look like `Version` is being set

File: http/server.go [hunk 3]
```diff
@@ -440,6 +440,33 @@ func (s *Server) handleExec(ctx *fasthttp.RequestCtx) {
 	})
 }
 
+func (s *Server) handleVersion(ctx *fasthttp.RequestCtx) {
+	if !ctx.IsPost() { // ctx.PostBody() blocks on GET
+		ctx.Error("unsupported method", http.StatusMethodNotAllowed)
+	}
+
+	requestInner := &pb.VersionRequest{}
```
Comment (talIguaz) [inline http/server.go:448]: you expect an empty request, why parse it and all that?

File: types.go [hunk 4]
```diff
@@ -343,6 +343,12 @@ const (
 	FailOnError = pb.ErrorOptions_FAIL
 )
 
+type VersionRequest struct {
```
Comment (talIguaz) [inline types.go:346]: doesn't look like you need any of that

File: http/server.go [hunk 5]
```diff
@@ -61,7 +61,7 @@ type Server struct {
 }
 
 // NewServer creates a new server
-func NewServer(config *frames.Config, addr string, logger logger.Logger, historyServer *utils.HistoryServer) (*Server, error) {
+func NewServer(config *frames.Config, addr string, logger logger.Logger, historyServer *utils.HistoryServer, version string) (*Server, error) {
```
Comment (talIguaz) [inline http/server.go:64]: If you need to have this in both "servers" maybe you can just have this statically here, meaning just dont go through the `api` when requesting version

File: grpc/server.go [hunk 6]
```diff
@@ -348,3 +348,17 @@ func (s *Server) History(request *pb.HistoryRequest, stream pb.Frames_HistorySer
 
 	return apiError
 }
+
+// Exec executes a command
+func (s *Server) Version(ctx context.Context, req *pb.VersionRequest) (*pb.VersionResponse, error) {
+	request := frames.VersionRequest{}
+
+	version, err := s.api.Version(&request)
```
Comment (talIguaz) [inline grpc/server.go:356]: again, no need to pass anything to the `s.api.Version`
Or alternatively just return `s.version`

File: clients/py/v3io_frames/grpc.py [hunk 7]
```diff
@@ -232,6 +238,17 @@ def do_history(self, backend, container, table, user, action, min_start_time, ma
         for frame in stub.History(request):
             yield msg2df(frame, self.frame_factory)
 
+    @grpc_raise(VersionError)
+    def _check_version(self):
+        stub = fgrpc.FramesStub(self._channel)
+        request = fpb.VersionRequest()
+        resp = stub.Version(request)
+        if resp.version:
```
Comment (talIguaz) [inline clients/py/v3io_frames/grpc.py:246]: is it ok to not get a version?

File: clients/py/v3io_frames/grpc.py [hunk 8]
```diff
@@ -232,6 +238,17 @@ def do_history(self, backend, container, table, user, action, min_start_time, ma
...
+            if __version__ != resp.version:
+                warnings.warn("Warning - Server version \'" + resp.version + "\' is different from client version \'" + __version__ + "\'. Some operations may not work.")
+            else:
```
Comment (talIguaz) [inline clients/py/v3io_frames/grpc.py:249]: the else should be one indent back :)

File: clients/py/v3io_frames/grpc.py [hunk 9]
```diff
@@ -38,6 +40,8 @@
     ('grpc.max_receive_message_length', GRPC_MESSAGE_SIZE),
 ]
 
+warnings.formatwarning = lambda msg, *args, **kwargs: f'{msg}\n'
```
Comment (talIguaz) [inline clients/py/v3io_frames/grpc.py:43]: why are you purposely ignoring args and kwargs?

File: clients/py/v3io_frames/grpc.py [hunk 10]
```diff
@@ -232,6 +238,17 @@ def do_history(self, backend, container, table, user, action, min_start_time, ma
...
             yield msg2df(frame, self.frame_factory)
 
+    @grpc_raise(VersionError)
+    def _check_version(self):
+        stub = fgrpc.FramesStub(self._channel)
+        request = fpb.VersionRequest()
+        resp = stub.Version(request)
+        if resp.version:
+            if __version__ != resp.version:
+                warnings.warn("Warning - Server version \'" + resp.version + "\' is different from client version \'" + __version__ + "\'. Some operations may not work.")
```
Comment (talIguaz) [inline clients/py/v3io_frames/grpc.py:248]: "Some operations may not work as expected" ?

File: clients/py/v3io_frames/http.py [hunk 11]
```diff
@@ -224,6 +230,30 @@ def _history(self, backend, container, table, user, action, min_start_time, max_
...
+        except json.JSONDecodeError as err:
+            raise VersionError(str(err))
+
+        version = out.get('version')
+        if not version:
+            return
+
+        if __version__ != version:
+            warnings.warn("Warning - Server version \'" + version + "\' is different from client version \'" + __version__ + "\'. Some operations may not work.")
+        else:
```
Comment (talIguaz) [inline clients/py/v3io_frames/http.py:254]: Looks like this should should be in the `if not version`. no?

File: cmd/framesd/framesd.go (hunk 2 above)
Comment (dinal) [reply cmd/framesd/framesd.go:91]: it does, by the build process
see in makefile -> FRAMES_BUILD_COMMAND

File: clients/py/v3io_frames/grpc.py (hunk 9 above)
Comment (dinal) [reply clients/py/v3io_frames/grpc.py:43]: warnings prints the msg in a very ugly format - 
`/Users/dinaleventol/PycharmProjects/framestest/v3io_frames_local/http.py:254: UserWarning: Warning - Server version 'unknown' is different from client version '0.4.1'. Some operations may not work.
  __version__ + "\'. Some operations may not work.")` 

googled it a bit and found this way of stripping the extra info warnings prints and keeping just the warn msg itself

**Review by talIguaz — APPROVED** [review]

File: Makefile [hunk 12]
```diff
@@ -45,7 +45,7 @@ grpc-go:
 .PHONY: grpc-py
 grpc-py:
 	cd clients/py && \
-	pipenv run python -m grpc_tools.protoc \
+	FRAMES_PYTHON_DEBUG=true pipenv run python -m grpc_tools.protoc \
```
Comment (talIguaz) [inline Makefile:48]: Do we need this flag here? it's just building the proto

File: clients/py/v3io_frames/grpc.py [hunk 13]
```diff
@@ -87,6 +92,9 @@ def __init__(self, *args, **kwargs):
         # create the session object, persist it between requests
         self._open_new_channel()
 
+        if self._channel and not os.environ.get("FRAMES_PYTHON_DEBUG"):
+            self._check_version()
```
Comment (talIguaz) [inline clients/py/v3io_frames/grpc.py:96]: better put this check in `__init__/Client` which would catch grpc and http

**Review by talIguaz — APPROVED** [review]

**Pull request merged**